src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,393 @@
+import { ShaderGraphNode, PORT_TYPE_IN, PORT_TYPE_SWITCH, comp2Type } from './shader-graph-node.js';
+
+import { Texture } from '../../graphics/texture.js';
+
+import { PIXELFORMAT_R8_G8_B8_A8, TEXTURETYPE_RGBM } from '../../graphics/graphics.js';
+
+var id = 0;
+
+var _placeholderTex;
+
+/**
+ * @private
+ * @class
+ * @name pc.ShaderGraphBuilder
+ * @classdesc A Shader Graph Builder class.
+ * @param {pc.Application} app - Application with shader graph core nodes registered that will be used to
+ * build the shader graph.
+ */
+var ShaderGraphBuilder = function (app) {
+    id++;
+
+    this.app = app;
+
+    this._graph = new ShaderGraphNode();
+    this._graph.name = 'graphRoot_' + id;
+
+    // for validation
+    this._addedNamedInputs = {};
+    this._addedNodes = [];
+    this._addedOutputs = {};
+
+    if (!_placeholderTex) {
+        // create texture
+        _placeholderTex = new Texture(this.app.graphicsDevice, {
+            width: 2,
+            height: 2,
+            format: PIXELFORMAT_R8_G8_B8_A8
+        });
+        _placeholderTex.name = 'placeholder';
+
+        // fill pixels with grey
+        var pixels = _placeholderTex.lock();
+        for (var i = 0; i < 4; i++) {
+            for (var c = 0; c < 4; c++) {
+                pixels[i * 4 + c] = [128, 128, 128, 255];
+            }
+        }

```

Comment:
This does not fill the texture with those pixels .. you're assigning an array of 4 elements to each pixel, that does do what you need. The final pixels array is just full of zeros.
Maybe look .fill, as you get typedArray from .lock()
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/fill

Reply:
Ah yes - made a mistake in adjusting this code when copying from the asset handler - I will switch to fill - thanks!